[scroll-animations] Refactor animationRange to be on Animation instead of StyleRareNonInheritedData

nmoucht opened the pull request:
#### e4d5669b38e02adac432d98a962d89cc3c68da89
<pre>
[scroll-animations] Refactor animationRange to be on Animation instead of StyleRareNonInheritedData
<a href="https://bugs.webkit.org/show_bug.cgi?id=281914">https://bugs.webkit.org/show_bug.cgi?id=281914</a>
<a href="https://rdar.apple.com/138426044">rdar://138426044</a>

Reviewed by Tim Nguyen.

Refactor animationRange to be on Animation, so that we can properly handle sequnces of animationRange
in a future patch, as well as update the underlying WebAnimation via CSSAnimation::syncPropertiesWithBackingAnimation.

* Source/WebCore/animation/TimelineRange.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapAnimationRangeStart const):
(WebCore::CSSToStyleMap::mapAnimationRangeEnd const):
* Source/WebCore/css/CSSToStyleMap.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForSingleAnimationRange):
(WebCore::valueForAnimationRange):
(WebCore::addValueForAnimationPropertyToList):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/platform/animation/Animation.cpp:
(WebCore::Animation::Animation):
(WebCore::Animation::animationsMatch const):
* Source/WebCore/platform/animation/Animation.h:
(WebCore::Animation::isRangeStartSet const):
(WebCore::Animation::isRangeEndSet const):
(WebCore::Animation::isEmpty const):
(WebCore::Animation::clearRangeStart):
(WebCore::Animation::clearRangeEnd):
(WebCore::Animation::clearAll):
(WebCore::Animation::rangeStart const):
(WebCore::Animation::rangeEnd const):
(WebCore::Animation::range const):
(WebCore::Animation::setRangeStart):
(WebCore::Animation::setRangeEnd):
(WebCore::Animation::setRange):
(WebCore::Animation::fillRangeStart):
(WebCore::Animation::fillRangeEnd):
(WebCore::Animation::isRangeStartFilled const):
(WebCore::Animation::isRangeEndFilled const):
(WebCore::Animation::isRangeFilled const):
(WebCore::Animation::initialRange):
(WebCore::Animation::initialRangeStart):
(WebCore::Animation::initialRangeEnd):
* Source/WebCore/platform/animation/AnimationList.cpp:
(WebCore::AnimationList::fillUnsetProperties):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTimelineScope):
(WebCore::RenderStyle::timelineScope const):
(WebCore::RenderStyle::initialAnimationRangeStart): Deleted.
(WebCore::RenderStyle::initialAnimationRangeEnd): Deleted.
(WebCore::RenderStyle::animationRangeStart const): Deleted.
(WebCore::RenderStyle::animationRangeEnd const): Deleted.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setTimelineScope):
(WebCore::RenderStyle::setAnimationRangeStart): Deleted.
(WebCore::RenderStyle::setAnimationRangeEnd): Deleted.
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:

Canonical link: <a href="https://commits.webkit.org/285609@main">https://commits.webkit.org/285609@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20dfd75abbf7a2feebb91447645b569940db055a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73168 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25976 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77392 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24406 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75283 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60402 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/379 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57497 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15974 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76235 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47515 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62964 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37916 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44154 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20442 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22735 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66008 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20796 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79050 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/61 "Found 1 new test failure: http/tests/ssl/applepay/ApplePayPaymentDetailsModifier.https.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65931 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62973 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-layout-tests-in-stress-mode") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65215 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7207 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/447 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3184 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/476 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/461 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/478 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1751,7 +1781,16 @@ static void addValueForAnimationPropertyToList(const RenderStyle& style, CSSValu
         } else
             list.append(valueForAnimationTimingFunction(CubicBezierTimingFunction::defaultTimingFunction()));
-    } else
+    } else if (property == CSSPropertyAnimationRangeStart) {
```
Can we change this to a switch statement?

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -1557,11 +1557,13 @@
             }
         },
         "animation-range": {
```
This should be set a longhand of `animation`, where we reset `animation-range` to the default value

nmoucht replied:
This should probably be a separate patch. I think animation-timeline also needs a similar fix.

graouts replied:
The `animation-timeline` property is not part of the `animation` shorthand (see [bug 279732](https://bugs.webkit.org/show_bug.cgi?id=279732)) and I don't believe `animation-range` is either.

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1708,51 +1708,110 @@ static Ref<CSSValue> valueForAnimationTimingFunction(const TimingFunction& timin
         if (!animation || !animation->isAllowsDiscreteTransitionsFilled())
             list.append(valueForTransitionBehavior(animation ? animation->allowsDiscreteTransitions() : Animation::initialAllowsDiscreteTransitions()));
-    } else if (property == CSSPropertyAnimationDuration || property == CSSPropertyTransitionDuration) {
+        return;
```
Please use `break;` and be consistent about having a newline after the `break;`

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/platform/animation/Animation.cpp:
```diff
@@ -135,7 +146,9 @@ bool Animation::animationsMatch(const Animation& other, bool matchProperties) co
         && m_compositeOperationSet == other.m_compositeOperationSet
         && m_allowsDiscreteTransitionsSet == other.m_allowsDiscreteTransitionsSet
-        && m_isNone == other.m_isNone;
+        && m_isNone == other.m_isNone
```
none should probably remain the last check (or moved to first).

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/platform/animation/Animation.h:
```diff
@@ -53,6 +53,9 @@ class Animation : public RefCounted<Animation> {
     bool isAllowsDiscreteTransitionsSet() const { return m_allowsDiscreteTransitionsSet; }
+    bool isRangeStartSet() const { return m_rangeStartSet; }
+    bool isRangeEndSet() const { return m_rangeEndSet; }
+
```
nit: extraneous new line

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSToStyleMap.cpp:
```diff
@@ -634,4 +634,18 @@ void CSSToStyleMap::mapNinePieceImageRepeat(const CSSValue& value, NinePieceImag
     image.setVerticalRule(fromCSSValue<NinePieceImageRule>(value.second()));
 }
 
+void CSSToStyleMap::mapAnimationRangeStart(Animation& animation, const CSSValue& value) const
```
can you move the new methods with the other animation methods?

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSToStyleMap.cpp:
```diff
@@ -634,4 +634,18 @@ void CSSToStyleMap::mapNinePieceImageRepeat(const CSSValue& value, NinePieceImag
+void CSSToStyleMap::mapAnimationRangeStart(Animation& animation, const CSSValue& value) const
+{
+    if (treatAsInitialValue(value, CSSPropertyAnimationRangeStart))
+        animation.setRangeStart({ });
```
```suggestion
        animation.setRangeStart(Animation::initialRangeStart());
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSToStyleMap.cpp:
```diff
@@ -634,4 +634,18 @@ void CSSToStyleMap::mapNinePieceImageRepeat(const CSSValue& value, NinePieceImag
+void CSSToStyleMap::mapAnimationRangeEnd(Animation& animation, const CSSValue& value) const
+{
+    if (treatAsInitialValue(value, CSSPropertyAnimationRangeEnd))
+        animation.setRangeEnd({ });
```
```suggestion
        animation.setRangeEnd(Animation::initialRangeEnd());
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/platform/animation/AnimationList.cpp:
```diff
@@ -60,6 +60,9 @@ void AnimationList::fillUnsetProperties()
     FILL_UNSET_PROPERTY(isAllowsDiscreteTransitionsSet, allowsDiscreteTransitions, fillAllowsDiscreteTransitions);
+    FILL_UNSET_PROPERTY(isRangeStartSet, rangeStart, fillRangeStart);
+    FILL_UNSET_PROPERTY(isRangeEndSet, rangeEnd, fillRangeEnd);
+
```
nit: remove extra new line

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1708,51 +1708,110 @@ static Ref<CSSValue> valueForAnimationTimingFunction(const TimingFunction& timin
             list.append(valueForTransitionBehavior(animation ? animation->allowsDiscreteTransitions() : Animation::initialAllowsDiscreteTransitions()));
-    } else if (property == CSSPropertyAnimationDuration || property == CSSPropertyTransitionDuration) {
+        break;
+    case CSSPropertyAnimationDuration:
+    case CSSPropertyTransitionDuration:
         if (!animation || !animation->isDurationFilled())
             list.append(valueForAnimationDuration(animation ? animation->duration() : Animation::initialDuration(), animation, animationList));
-    } else if (property == CSSPropertyAnimationDelay || property == CSSPropertyTransitionDelay) {
+        break;
```
Please be consistent about whether you add a newline after the break or not. I have a slight preference for adding one, but please do so after the other cases as well.

nmoucht replied:
I'm not sure what you mean this switch is consistent with not adding newlines after breaks. Other switches in this file don't add a newline so I'm going follow that.

nt1m replied:
Ah nevermind, Github is playing tricks when showing the diff

nmoucht committed:
[scroll-animations] Refactor animationRange to be on Animation instead of StyleRareNonInheritedData
https://bugs.webkit.org/show_bug.cgi?id=281914
rdar://138426044

Reviewed by Tim Nguyen.

Refactor animationRange to be on Animation, so that we can properly handle sequnces of animationRange
in a future patch, as well as update the underlying WebAnimation via CSSAnimation::syncPropertiesWithBackingAnimation.

* Source/WebCore/animation/TimelineRange.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapAnimationRangeStart const):
(WebCore::CSSToStyleMap::mapAnimationRangeEnd const):
* Source/WebCore/css/CSSToStyleMap.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForSingleAnimationRange):
(WebCore::valueForAnimationRange):
(WebCore::addValueForAnimationPropertyToList):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/platform/animation/Animation.cpp:
(WebCore::Animation::Animation):
(WebCore::Animation::animationsMatch const):
* Source/WebCore/platform/animation/Animation.h:
(WebCore::Animation::isRangeStartSet const):
(WebCore::Animation::isRangeEndSet const):
(WebCore::Animation::isEmpty const):
(WebCore::Animation::clearRangeStart):
(WebCore::Animation::clearRangeEnd):
(WebCore::Animation::clearAll):
(WebCore::Animation::rangeStart const):
(WebCore::Animation::rangeEnd const):
(WebCore::Animation::range const):
(WebCore::Animation::setRangeStart):
(WebCore::Animation::setRangeEnd):
(WebCore::Animation::setRange):
(WebCore::Animation::fillRangeStart):
(WebCore::Animation::fillRangeEnd):
(WebCore::Animation::isRangeStartFilled const):
(WebCore::Animation::isRangeEndFilled const):
(WebCore::Animation::isRangeFilled const):
(WebCore::Animation::initialRange):
(WebCore::Animation::initialRangeStart):
(WebCore::Animation::initialRangeEnd):
* Source/WebCore/platform/animation/AnimationList.cpp:
(WebCore::AnimationList::fillUnsetProperties):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTimelineScope):
(WebCore::RenderStyle::timelineScope const):
(WebCore::RenderStyle::initialAnimationRangeStart): Deleted.
(WebCore::RenderStyle::initialAnimationRangeEnd): Deleted.
(WebCore::RenderStyle::animationRangeStart const): Deleted.
(WebCore::RenderStyle::animationRangeEnd const): Deleted.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setTimelineScope):
(WebCore::RenderStyle::setAnimationRangeStart): Deleted.
(WebCore::RenderStyle::setAnimationRangeEnd): Deleted.
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:

Canonical link: https://commits.webkit.org/285609@main
